specs/core/0_beacon-chain.md

```diff
@@ -2002,80 +1993,49 @@ def apply_rewards(state: BeaconState) -> None:
-    Note that this function mutates ``state``.
-    """
-    current_epoch = get_current_epoch(state)
-    # The active validators
-    active_validator_indices = get_active_validator_indices(state.validator_registry, current_epoch)
-    # The total effective balance of active validators
-    total_balance = get_total_balance(state, active_validator_indices)
-
-    # The maximum balance churn in Gwei (for deposits and exits separately)
-    max_balance_churn = max(
-        MAX_DEPOSIT_AMOUNT,
-        total_balance // (2 * MAX_BALANCE_CHURN_QUOTIENT)
-    )
-
-    # Activate validators within the allowable balance churn
-    balance_churn = 0
     for index, validator in enumerate(state.validator_registry):
-        if validator.activation_epoch == FAR_FUTURE_EPOCH and get_balance(state, index) >= MAX_DEPOSIT_AMOUNT:
-            # Check the balance churn would be within the allowance
-            balance_churn += get_effective_balance(state, index)
-            if balance_churn > max_balance_churn:
-                break
-
-            # Activate validator
-            activate_validator(state, index, is_genesis=False)
-
-    # Exit validators within the allowable balance churn
-    if current_epoch < state.validator_registry_update_epoch + LATEST_SLASHED_EXIT_LENGTH:
-        balance_churn = (
-            state.latest_slashed_balances[state.validator_registry_update_epoch % LATEST_SLASHED_EXIT_LENGTH] -
-            state.latest_slashed_balances[current_epoch % LATEST_SLASHED_EXIT_LENGTH]
-        )
-
-        for index, validator in enumerate(state.validator_registry):
-            if validator.exit_epoch == FAR_FUTURE_EPOCH and validator.initiated_exit:
-                # Check the balance churn would be within the allowance
-                balance_churn += get_effective_balance(state, index)
-                if balance_churn > max_balance_churn:
-                    break
+        balance = get_balance(state, index)
+        if validator.activation_eligibility_epoch == FAR_FUTURE_EPOCH and balance >= MAX_DEPOSIT_AMOUNT:
+            state.activation_eligibility_epoch = get_current_epoch(state)
 
-                # Exit validator
-                exit_validator(state, index)
-
-    state.validator_registry_update_epoch = current_epoch
+        if is_active_validator(validator, get_current_epoch(state)) and balance < EJECTION_BALANCE:
+            initiate_validator_exit(state, index)
 ```
 
+#### Validator registry and start shard
+
 Run the following function:
 
 ```python
 def update_registry(state: BeaconState) -> None:
+    current_epoch = get_current_epoch(state)
     # Check if we should update, and if so, update
-    if state.finalized_epoch > state.validator_registry_update_epoch:
-        update_validator_registry(state)
+
+    activations_since_finalization = len([index in state.validator_registry if
+        state.validator_registry[index].activation_epoch > state.finalized_epoch + ACTIVATION_EXIT_DELAY
+    ])
```

Comment:
I think we can remove this statement and the if statement entirely, because the line of code `for index in indices_for_activation[:MAX_EXIT_DEQUEUES_PER_EPOCH - activations_since_finalization]:` does that function already.

Reply:
Reworked, saving a bunch of lines :+1:.

Why does `get_delayed_activation_exit_epoch` have a `+ 1`? cc @djrtwo

Reply:
It's `epoch + 1 + ACTIVATION_EXIT_DELAY` because `ACTIVATION_EXIT_DELAY` is the _minimum_ activation delay, so if you're currently inside epoch n, to ensure a delay of four full epochs you want to return n+5 so that you get the entire span of epochs [n+1 .... n+4] in between.